---
title: "Analysis_OleateMicrobiota"
output: html_document
---
# Project Description  
In this current project, we have 3 groups of mice. A control group that was fed 10% fat diet, an oleate group which was fed with an oleate diet and a HFD group which was fed with 60% fat diet. The project included 19 mice. We collected feces, contents from the ileum, cecum and colon and we performed 16S rRNA sequencing on them. The DNA extraction and sequencing was done in the university of Michigan Microbiome core. The details of sequencing procedures are present in the `Microbiome_core_sequencing_details.xlsx` excel file.  
  
# Preprocessing and classifying OTUs in mothur  
First, we created a file that assign the fastq files to the samples. The file is called `oleate.files` in the data directory.  
We used mothur to preprocess the data and to clasify OTUs. An executable version of mothur for windows (version 1.44.3) was installed in the folder containing the fastq files. The original fastq files were processed using mothur pipeline using the following code:   
```
# setting the input and output directories  
set.dir(input=Data, output= ButyrateResults)
# making contigs  
make.contigs(file=oleate.files, processors=4)
screen.seqs(fasta=current, group=current, maxambig=0, maxlength=275)
unique.seqs(fasta=current)
count.seqs(name=current, group=current)
summary.seqs(count=current)
# make aan align reference file from silva original reference file. The silva reference files (version 138.1) were downloaded from <https://mothur.org/wiki/silva_reference_files/>
pcr.seqs(fasta=silva.nr_v138_1.align, start=11894, end=25319, keepdots=F, processors=8)
# align sequences  
align.seqs(fasta= oleate.trim.contigs.good.unique.fasta, reference= silva.nr_v138_1.pcr.align)
summary.seqs(fasta= current, count= current)
screen.seqs(fasta=current, count=current, summary=current, start=1968, end=11550, maxhomop=8)
summary.seqs(fasta= current, count= current)
filter.seqs(fasta=current, vertical=T, trump=.)
unique.seqs(fasta=current, count=current)
pre.cluster(fasta=current, count=current, diffs=2)
chimera.vsearch(fasta=current, count=current, dereplicate=t)
remove.seqs(fasta=current, accnos=current)
summary.seqs(fasta=current, count=current)
# classify sequences
classify.seqs(fasta=current, count=current, reference= silva.nr_v138_1.pcr.align, taxonomy= silva.nr_v138_1.tax, cutoff=80)
remove.lineage(fasta=current, count=current, taxonomy=current, taxon=Chloroplast-Mitochondria-unknown-Archaea-Eukaryota)
summary.tax(taxonomy=current, count=current)
rename.file(taxonomy=oleate.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.opti_mcc.0.03.cons.taxonomy, shared= oleate.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.opti_mcc.shared)
# investigate the read counts in each sample and selecting the least count for subsampling
count.groups(shared=oleate.opti_mcc.shared)
sub.sample(shared= current, size= 2766)
rarefaction.single(shared=current, calc=sobs, freq=100)
summary.single(shared=current, calc=nseqs-coverage-sobs-invsimpson, subsample=T)
# calculating distances and principal coordinate analysis
dist.shared(shared=current, calc=thetayc-jclass, subsample=t)
pcoa(phylip=current)
nmds(phylip=current)
nmds(phylip=current, mindim=3, maxdim=3)
# we created a set of txt files in excel to reflect the design and metadata of the samples. these files are present in the data directory under the names `condition.design`, `site.design` and `oleate.metadata`. Next, we proceeded with performing AMOVA to detect significance
amova(phylip=current, design=condition.design)
amova(phylip=current, design=site.design)
homova(phylip=current, design=condition.design)
homova(phylip=current, design=site.design)
```
# Further analysis and visualization in R  
We created an R project and copied the following files to the project directory:  
1- oleate.opti_mcc.shared (the OTU table of all samples)  
2- oleate.opti_mcc.0.03.subsample.shared (the OTU table after subsampling)  
3-oleate.metadata (a txt file that includes the sample metadata)  
4- oleate.taxonomy (the taxonomy file that was generated by mothur. we modified this file in excel. As it is from mothur, there are not column headers for order, family genus etc. in the cons.taxonomy file and this must be fixed first. This is how I did it: read cons.taxonomy file into excel and choose semicolon as a separator so each tax level is therefore in its own column. then delete header 'taxonomy' and put in appropriate header for each column (order or family or genus etc. Then select all, press ctrl+shift+H on mac to find and replace. Replate all `(*)` with nothing)   
## Install and load required libraries  
```{r, message=FALSE}
library("phyloseq")
library("ggplot2")
library("plyr")
library("vegan")
library("grid")
library("directlabels")
library("knitr")
library("clustsig")
library("ellipse")
library("ape")
library("DESeq2")
library("microbial")
library("VennDiagram")
library("microbiome")
library("microbiomeutilities")
library("viridis")
library("RColorBrewer")
library("ggpubr")
library("patchwork")
```
## Make a phyloseq object from mothur files  
```{r}  
## import data 
sharedFile = read.table("oleate.opti_mcc.shared")
sharedFile = t(sharedFile) ## transform data
rownames(sharedFile) = sharedFile[,1]  ## define rowname
colnames(sharedFile) = sharedFile[2,]  ## define column names
sharedFile <-as.matrix(sharedFile)
sharedFile = sharedFile[,2:57]       ## only include columns that reflects OTU numbers (remember order [row,colum]) 2 -> 57 for me
## as I had 57 samples
sharedFile = sharedFile[4:819,]     ## and which rows you want. I have 819 OTUs so I changed this number to 819
# original => sharedFile = sharedFile[4:819,]
class(sharedFile) <- "numeric"
head(sharedFile)  ## look at head first 7 lines to see it's made correclty
## Import subsampled otu matrix (26880 seqs)
sharedsubFile = read.table('oleate.opti_mcc.0.03.subsample.shared')
sharedsubFile = t(sharedsubFile)
rownames(sharedsubFile) = sharedsubFile[,1]
colnames(sharedsubFile) = sharedsubFile[2,]
sharedsubFile = sharedsubFile[,2:57]
sharedsubFile = sharedsubFile[4:494,]
class(sharedsubFile) <- "numeric"
head(sharedsubFile)
sharedsubFile <-as.matrix(sharedsubFile)
## Import taxonomy file 
## Copy and paste into notepad and save. Then carry on: 
taxFile = read.table('oleate.taxonomy', header=T, sep='\t')
rownames(taxFile) = taxFile[,1]
taxFile = taxFile[,3:8]
taxFile = as.matrix(taxFile)
head(taxFile)
## import metadata file
metaFile = read.table('oleate.metadata', header=T, sep='\t')
rownames(metaFile) = metaFile[,1]
metaFile = metaFile[,1:4]
head(metaFile)
## Create phyloseq object
OTU = otu_table(sharedFile, taxa_are_rows = TRUE)
OTUsub = otu_table(sharedsubFile, taxa_are_rows = TRUE)
TAX = tax_table(taxFile)
META = sample_data(metaFile)
physeq = phyloseq(OTU, TAX, META)
physeqSub = phyloseq(OTUsub, TAX, META)
## Get rid of any OTUs not present in any samples and get relative abundance
microSub <- prune_taxa(taxa_sums(physeqSub) > 0, physeqSub)
microSubRel = transform_sample_counts(microSub, function(x) x / sum(x) )
microSubRelFilt = filter_taxa(microSubRel, function(x) mean(x) > 1e-5, TRUE)
# for subsampled shared file
# create a tree and a new phyloseq object
random_tree = rtree(ntaxa(microSub), rooted=TRUE, tip.label=taxa_names(microSub))
plot(random_tree)
microSubt = merge_phyloseq(microSub, random_tree)
```

## Plotting core graphs
```{r, warning=FALSE}
# Plotting rarefaction
rarecurve(t(otu_table(microSubt)), step=50, cex=0.5)
# Plot a prettier rarefaction curve
# set seed
set.seed(1)
subsamples <- seq(0, 5000, by=100)[-1]
#subsamples = c(10, 5000, 10000, 20000, 30000)
p <- plot_alpha_rcurve(microSub, index="observed", 
                       subsamples = subsamples,
                       lower.conf = 0.025, 
                       upper.conf = 0.975,
                       group="condition",
                       label.color = "brown3",
                       label.size = 3,
                       label.min = TRUE) 
# change color of line 
mycols <- c("brown3", "steelblue","grey50")

p <- p + scale_color_manual(values = mycols) + 
  scale_fill_manual(values = mycols)
print(p)
# Plotting taxonomy
plot_bar(microSubt, fill="Phylum") + 
  facet_wrap(~condition, scales = "free_x", nrow = 1) +
  theme(text = element_text(size=16))+geom_bar(stat="identity") +
  scale_fill_brewer(type = "seq", palette = "Spectral") +
  theme(
    legend.background = element_rect(
      fill = "lemonchiffon", 
      colour = "grey50", 
      size = 1
    )
  )
plot_bar(microSubt, fill="Phylum") + 
  facet_wrap(~site, scales = "free_x", nrow = 1) +
  theme(text = element_text(size=16))+geom_bar(stat="identity") +
  scale_fill_brewer(type = "seq", palette = "Spectral") +
  theme(
    legend.background = element_rect(
      fill = "lemonchiffon", 
      colour = "grey50", 
      size = 1
    )
  ) 
plot_bar(microSub, fill="Phylum") +
  facet_grid(condition~site, scales = "free") +  ## separate facets
  theme(text = element_text(size=16)) +  ## increase font of all elements
  geom_bar(stat="identity")+
  scale_fill_brewer(type = "seq", palette = "Spectral") +
  theme(
    legend.background = element_rect(
      fill = "lemonchiffon", 
      colour = "grey50", 
      size = 1
    )
  )
# Plotting PCoA
my.PCoA2 <- ordinate(microSub, "PCoA", "bray")
plot_ordination(microSub, my.PCoA2, type = "group", color = "condition", shape = "site")+theme(text = element_text(size=20))+ geom_point(size=5) + scale_color_manual(values=c("brown3", "steelblue","grey50"))
plot_ordination(microSub, my.PCoA2, type = "group", color = "condition")+ facet_wrap(~site, scales = "free_x", nrow = 1) + geom_point(size=5)+theme(text = element_text(size=20)) + scale_color_manual(values=c("brown3", "steelblue","grey50"))
plot_ordination(microSub, my.PCoA2, type = "group", color = "condition", shape = "site")+theme(text = element_text(size=20))+ geom_point(size=5)+ 
  geom_line() + scale_color_manual(values=c("brown3", "steelblue","grey50"))
plot_ordination(microSub, my.PCoA2, type = "group", color = "condition")+theme(text = element_text(size=20))+ geom_point(size=5)+ stat_ellipse(level=0.9)  + scale_color_manual(values=c("brown3", "steelblue","grey50"))
# PCoA plot using the unweighted UniFrac as distance
wunifrac_dist = phyloseq::distance(microSubt, method="unifrac", weighted=F)
ordination = ordinate(microSubt, method="PCoA", distance=wunifrac_dist)
plot_ordination(microSubt, ordination, color="condition") + theme(aspect.ratio=1)+
  ggtitle("PCoA: unweigthed Unifrac")+geom_point(size=5)  + scale_color_manual(values=c("brown3", "steelblue","grey50"))
#Plot PCoA using the weighted UniFrac as distance
wunifrac_dist = phyloseq::distance(microSubt, method="unifrac", weighted=T)
ordUF = ordinate(microSubt, method="PCoA", distance=wunifrac_dist)
plot_ordination(microSubt, ordUF, color = "condition") + 
  ggtitle("PCoA: weigthed Unifrac")+geom_point(size=5) + scale_color_manual(values=c("brown3", "steelblue","grey50"))
```  

## Normalization and plotting figures by group  
```{r}
# Normalization
phy <- normalize(microSubt, method = "relative")
# Plot taxonomy by group
plotbar(phy,level="Phylum", group="condition") + theme(axis.text.x = element_text(size = 16))
plotbar(phy,level="Phylum", group="site") + theme(axis.text.x = element_text(size = 16))
# Plot alpha diversity
plotalpha(microSubt, group = "condition", color = c("brown3","grey50", "steelblue"))
plotalpha(microSubt, group = "site")
# Test for significance between groups
beta_condition <-betatest(phy,group="condition")
beta_site <-betatest(phy,group="site")
# Biomarkers discovery and plotting
bio <- biomarker(microSubt,group="condition")
plotmarker(bio,level="Genus")
# LefSe testing and plotting
lda <- ldamarker(microSubt,group="condition")
write.csv(lda, "lda.csv", row.names = FALSE)
# open the csv file in excel, separate the taxonomy to kingdom, phylum,... and save it as lda2.csv so that the names can appear better on the graph
lda2 <- read.csv(file = 'lda2.csv')
plotLDA(lda2,group=c("10%Fat","60%Fat"),lda=5,pvalue=0.05)+theme(axis.text.y = element_text(size = 16))+theme(text = element_text(size=20))
plotLDA(lda2,group=c("10%Fat","Oleate"),lda=5,pvalue=0.05)+theme(axis.text.y = element_text(size = 16))+theme(text = element_text(size=20))
plotLDA(lda2,group=c("Oleate","60%Fat"),lda=5,pvalue=0.05) +theme(axis.text.y = element_text(size = 16))+theme(text = element_text(size=20))
```

## Modified difftest function from microbial package to test for significance
```{r, message=FALSE}
# create difftest2 function
difftest2<-function(physeq,group,pvalue=0.05,padj=NULL,log2FC=0,gm_mean=TRUE,fitType="local",quiet=FALSE){
  if(!taxa_are_rows(physeq)){
    physeq<-t(physeq)
  }
  otu <- as(otu_table(physeq),"matrix")
  tax <- as.data.frame(as.matrix(tax_table(physeq)))
  colData<-as(sample_data(physeq),"data.frame")
  colData$condition<-colData[,group]
  contrasts<-levels(factor(unique(colData$condition)))[1:2]
  if(isTRUE(gm_mean)){
    countData<-round(otu, digits = 0)
  }else{
    countData<-round(otu, digits = 0)+1
  }
  dds <- DESeqDataSetFromMatrix(countData, colData, as.formula(~condition))
  if(isTRUE(gm_mean)){
    geoMeans = apply(counts(dds), 1, gm_mean)
    dds = estimateSizeFactors(dds, geoMeans = geoMeans)
  }
  dds <- DESeq(dds, fitType=fitType)
  res <- results(dds,contrast=c("condition",contrasts),cooksCutoff = FALSE)
  res_tax = cbind(as.data.frame(res), as.matrix(countData[rownames(res), ]))
  if(!is.null(padj)){
    pval<-padj
    sig <- rownames(subset(res,padj<pval&abs(log2FoldChange)>log2FC))
  }else{
    pval<-pvalue
    sig <- rownames(subset(res,pvalue<pval&abs(log2FoldChange)>log2FC))
  }
  res_tax$Significant<- "No"
  res_tax$Significant <- ifelse(rownames(res_tax) %in% sig, "Yes", "No")
  res_tax <- cbind(res_tax, tax[rownames(res),])
  return(as.data.frame(res_tax))
}
# Test for significant bacteria using microbial package
mic_res <- difftest2(microSubt,group="condition", gm_mean=FALSE)
plotdiff(mic_res,level="Genus",padj=0.001,log2FC = 3,fontsize.y = 14)
```  

## Test for significant OTUs using DESeq2  
```{r}
sample_data(microSubt)$condition <- as.factor(sample_data(microSubt)$condition)
ds = phyloseq_to_deseq2(microSubt, ~ condition)
ds = DESeq(ds)
alpha = 0.01
# compare Oleate and 60%Fat
res = results(ds, contrast=c("condition", "Oleate", "60%Fat"), alpha=alpha)
res = res[order(res$padj, na.last=NA), ]
res_sig = res[(res$padj < alpha), ]
res_sig
res_sig = cbind(as(res_sig, "data.frame"), as(tax_table(microSub)[rownames(res_sig), ], "matrix"))
ggplot(res_sig, aes(x=Genus, y=log2FoldChange, color=Phylum)) +
  geom_jitter(size=5, width = 0.2) + theme(text = element_text(size=16)) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
# compare 10%Fat and 60%Fat
res2 = results(ds, contrast=c("condition", "10%Fat", "60%Fat"), alpha=alpha)
res2 = res2[order(res2$padj, na.last=NA), ]
res_sig2 = res2[(res2$padj < alpha), ]
res_sig2
res_sig2 = cbind(as(res_sig2, "data.frame"), as(tax_table(microSub)[rownames(res_sig2), ], "matrix"))
ggplot(res_sig2, aes(x=Genus, y=log2FoldChange, color=Phylum)) +
  geom_jitter(size=5, width = 0.2) + theme(text = element_text(size=16)) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
# compare 10%Fat and Oleate
res3 = results(ds, contrast=c("condition", "10%Fat", "Oleate"), alpha=alpha)
res3 = res3[order(res3$padj, na.last=NA), ]
res_sig3 = res3[(res3$padj < alpha), ]
res_sig3
res_sig3 = cbind(as(res_sig3, "data.frame"), as(tax_table(microSubt)[rownames(res_sig3), ], "matrix"))
ggplot(res_sig3, aes(x=Genus, y=log2FoldChange, color=Phylum)) +
  geom_jitter(size=5, width = 0.2) + theme(text = element_text(size=16)) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
```

## Plot heatmap  
```{r, results='hide'}
plot_taxa_heatmap(microSubt,
                  subset.top = 20,
                  VariableA = "condition",
                  heatcolors = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
                  transformation = "log10")
```  

## Venn Diagram
```{r}
pseq<-microSubt
table(meta(pseq)$condition, useNA = "always")
# convert to relative abundances
pseq.rel <- microbiome::transform(pseq, "compositional")
disease_states <- unique(as.character(meta(pseq.rel)$condition))
print(disease_states)
#Write a for loop to go through each of the disease_states one by one and combine identified core taxa into a list.
list_core <- c() # an empty object to store information

for (n in disease_states){ # for each variable n in DiseaseState
  #print(paste0("Identifying Core Taxa for ", n))
  
  ps.sub <- subset_samples(pseq.rel, condition == n) # Choose sample from DiseaseState by n
  
  core_m <- core_members(ps.sub, # ps.sub is phyloseq selected with only samples from g 
                         detection = 0.001, # 0.001 in atleast 90% samples 
                         prevalence = 0.75)
  print(paste0("No. of core taxa in ", n, " : ", length(core_m))) # print core taxa identified in each DiseaseState.
  list_core[[n]] <- core_m # add to a list core taxa for each group.
  #print(list_core)
}
print(list_core)
# Specify colors and plot venn
# supplying colors in the order they appear in list_core
mycols <- c("brown3", "grey50", "steelblue")
venn.diagram(list_core,fill = mycols, filename = 'venn_diagramm.png',
             output=TRUE)
```  
![Venn_Diagram](venn_diagramm.png)

## Plotting top 4 genus  
```{r, warning=FALSE}
pa <- aggregate_taxa(microSubt, "Genus")
top_four <- top_taxa(pa, 4)
top_four
mycols <- c("brown3", "steelblue", "grey50")
p <- plot_listed_taxa(pa, top_four, 
                      group= "condition",
                      group.order = c("10%Fat","Oleate","60%Fat"),
                      group.colors = mycols,
                      add.violin = TRUE,
                      violin.opacity = 0.3,
                      dot.opacity = 0.25,
                      box.opacity = 0.25,
                      panel.arrange= "wrap")
comps <- make_pairs(sample_data(pa)$condition)
p <- p + stat_compare_means(
  comparisons = comps,
  label = "p.format",
  tip.length = 0.05,
  method = "wilcox.test")
print(p + ylab("Relative abundance") + scale_y_continuous(labels = scales::percent))
```

## Get taxa summary by group(s)
```{r}
p0 <- microSubt
p0.gen <- aggregate_taxa(microSubt,"Genus")
x.d <- dominant_taxa(p0,level = "Genus", group="condition")
head(x.d$dominant_overview)
tx.sum1 <- taxa_summary(p0, "Phylum")
p0.rel <- microbiome::transform(p0, "compositional")

grp_abund <- get_group_abundances(p0.rel, 
                                  level = "Phylum", 
                                  group="condition",
                                  transform = "compositional")
mycols <- c("brown3", "steelblue","grey50")
# clean names 
grp_abund$OTUID <- gsub("p__", "",grp_abund$OTUID)
grp_abund$OTUID <- ifelse(grp_abund$OTUID == "", 
                          "Unclassified", grp_abund$OTUID)
mean.plot <- grp_abund %>% # input data
  ggplot(aes(x= reorder(OTUID, mean_abundance), # reroder based on mean abundance
             y= mean_abundance,
             fill=condition)) + # x and y axis 
  geom_bar(     stat = "identity", 
                position=position_dodge()) + 
  scale_fill_manual("condition", values=mycols) + # manually specify colors
  theme_bw() + # add a widely used ggplot2 theme
  ylab("Mean Relative Abundance") + # label y axis
  xlab("Phylum") + # label x axis
  coord_flip() # rotate plot 
mean.plot
```  

## Plotting density of reads per sample  
```{r}
p0 <- microSubt
print_ps(p0)
kable(head(tax_table(p0)))
# reduce size for example
ps0 <- core(p0, detection = 10, prevalence = 20 / 100)
# Add a prefix to taxa labels
ps0.f2 <- format_to_besthit(ps0, prefix = "MyBug-")
kable(head(tax_table(ps0.f2))[3:6])
p <- plot_read_distribution(ps0.f2, groups = "condition", 
                            plot.type = "density")
print(p + theme_biome_utils())
ps0 <- core(p0, detection = 10, prevalence = 20 / 100)
pseq_df <- phy_to_ldf(ps0, transform.counts = NULL)
kable(head(pseq_df))
# Plot Density of phyla
pseq <- microSubt

# check 10%Fat
control_ps <- subset_samples(pseq, condition=="10%Fat")
p_hc <- taxa_distribution(control_ps) + 
  theme_biome_utils() + 
  labs(title = "10%Fat")

# check Oleate
oleate_ps <- subset_samples(pseq, condition=="Oleate")
p_oleate <- taxa_distribution(oleate_ps) + 
  theme_biome_utils() + 
  labs(title = "Oleate")

# check 60%Fat
HFD_ps <- subset_samples(pseq, condition=="60%Fat")
p_hfd <- taxa_distribution(HFD_ps) + 
  theme_biome_utils() + 
  labs(title = "60%Fat")

# harnessing the power of patchwork
p_hc / p_oleate / p_hfd + plot_layout(guides = "collect") + 
  plot_annotation(tag_levels = "A")
```

## Filtering the dataset  
```{r}
# Select specific taxa
microFirm <- subset_taxa(microSubt, Phylum %in% "Firmicutes")
random_tree2 = rtree(ntaxa(microFirm), rooted=TRUE, tip.label=taxa_names(microFirm))
plot(random_tree2)
microFirm = merge_phyloseq(microFirm, random_tree2)
microFirm
# Subset samples by site or condition
microFecal <- subset_samples(microSubt, site=="fecal")
microFecal
# if you needed to remove candidate outliers, can use the below to remove sample e.g. f4406
microSubt.1 <- prune_samples(sample_names(microSubt) != "f4406", microSubt)
```  

## Convert Phyloseq object to a dataframe  
```{r}
x<-psmelt(microSubt)
head(x)
```  

# Session Info  
```{r}
sessionInfo()
```