---
title: "Identification of butyrate-producing bacteria"
output: html_document
---
We start with the `microSubt` phyloseq object created in the `Analysis-OleateMicrobiota` file. We create a `csv` file containing the major butyrate-producing microbiota reported in literature called `butyrate.csv`. This file was created from the silva v138.1 taxonomy file. Then, we filter the `microSubt` with genera from `butyrate.csv`.  
```{r, message=FALSE}
# load libraries
library("phyloseq")
library("ggplot2")
library("plyr")
library("vegan")
library("grid")
library("directlabels")
library("knitr")
library("clustsig")
library("ellipse")
library("ape")
library("DESeq2")
library("microbial")
library("VennDiagram")
library("microbiome")
library("microbiomeutilities")
library("viridis")
library("RColorBrewer")
library("ggpubr")
library("patchwork")
```
# Select Butyrate-forming bacteria from the microSubt phyloseq object  
```{r}
but <- read.csv("butyrate.csv", header = TRUE, sep = ",")
dim(but)
head(but$Genus)
```  

# Make a phyloseq object from mothur files  
```{r}  
## import data 
sharedFile = read.table("oleate.opti_mcc.shared")
sharedFile = t(sharedFile) ## transform data
rownames(sharedFile) = sharedFile[,1]  ## define rowname
colnames(sharedFile) = sharedFile[2,]  ## define column names
sharedFile <-as.matrix(sharedFile)
sharedFile = sharedFile[,2:57]       ## only include columns that reflects OTU numbers (remember order [row,colum]) 2 -> 57 for me
## as I had 57 samples
sharedFile = sharedFile[4:819,]     ## and which rows you want. I have 819 OTUs so I changed this number to 819
# original => sharedFile = sharedFile[4:819,]
class(sharedFile) <- "numeric"
head(sharedFile)  ## look at head first 7 lines to see it's made correclty
## Import subsampled otu matrix (26880 seqs)
sharedsubFile = read.table('oleate.opti_mcc.0.03.subsample.shared')
sharedsubFile = t(sharedsubFile)
rownames(sharedsubFile) = sharedsubFile[,1]
colnames(sharedsubFile) = sharedsubFile[2,]
sharedsubFile = sharedsubFile[,2:57]
sharedsubFile = sharedsubFile[4:494,]
class(sharedsubFile) <- "numeric"
head(sharedsubFile)
dim(sharedsubFile)
sharedsubFile <-as.matrix(sharedsubFile)
## Import taxonomy file 
## Copy and paste into notepad and save. Then carry on: 
taxFile = read.table('oleate.taxonomy', header=T, sep='\t')
rownames(taxFile) = taxFile[,1]
taxFile = taxFile[,3:8]
taxFile = as.matrix(taxFile)
head(taxFile)
taxBut <- as.data.frame(taxFile)
taxBut <-taxBut %>% 
  mutate(Family = ifelse(Family %in% but$Family, "Butyrate-producing bacteria", Family))
taxBut <- within(taxBut, Kingdom[Family == "Butyrate-producing bacteria"] <- '')
taxBut <- within(taxBut, Phylum[Family == "Butyrate-producing bacteria"] <- '')
taxBut <- within(taxBut, Order[Family == "Butyrate-producing bacteria"] <- '')
taxBut <- within(taxBut, Class[Family == "Butyrate-producing bacteria"] <- '')
taxBut <-as.matrix(taxBut)
## import metadata file
metaFile = read.table('oleate.metadata', header=T, sep='\t')
rownames(metaFile) = metaFile[,1]
metaFile = metaFile[,1:4]
head(metaFile)
## Create phyloseq object
OTU = otu_table(sharedFile, taxa_are_rows = TRUE)
OTUsub = otu_table(sharedsubFile, taxa_are_rows = TRUE)
TAX = tax_table(taxFile)
TaxBut=tax_table(taxBut)
META = sample_data(metaFile)
physeq = phyloseq(OTU, TAX, META)
physeqSub = phyloseq(OTUsub, TAX, META)
phytaxBut = phyloseq(OTUsub, TaxBut, META)
physeqSub
## Get rid of any OTUs not present in any samples and get relative abundance
microSub <- prune_taxa(taxa_sums(physeqSub) > 0, physeqSub)
microSubRel = transform_sample_counts(microSub, function(x) x / sum(x) )
microSubRelFilt = filter_taxa(microSubRel, function(x) mean(x) > 1e-5, TRUE)
microtaxBut <-prune_taxa(taxa_sums(phytaxBut) > 0, physeqSub)
# for subsampled shared file
# create a tree and a new phyloseq object
random_tree = rtree(ntaxa(microSub), rooted=TRUE, tip.label=taxa_names(microSub))
plot(random_tree)
microSubt = merge_phyloseq(microSub, random_tree)
random_tree2 = rtree(ntaxa(phytaxBut), rooted=TRUE, tip.label=taxa_names(phytaxBut))
phytaxbutT  = merge_phyloseq(microtaxBut, random_tree2)
# subset butyrate producing bacteria
microBut <- subset_taxa(microSubt, Genus %in% but$Genus)
microBut
```  

# Plotting core graphs
```{r}
# Plotting rarefaction
rarecurve(t(otu_table(microBut)), step=50, cex=0.5)
# Plotting taxonomy
plot_bar(microBut, fill="Family") + facet_wrap(~condition, scales = "free_x", nrow = 1)+theme(text = element_text(size=16))+geom_bar(stat="identity")+
  scale_fill_brewer(type = "seq", palette = "Spectral")+
  theme(
    legend.background = element_rect(
      fill = "lemonchiffon", 
      colour = "grey50", 
      size = 1
    )
  )
plot_bar(microBut, fill="Family") + facet_wrap(~site, scales = "free_x", nrow = 1)+theme(text = element_text(size=16))+geom_bar(stat="identity")+
  scale_fill_brewer(type = "seq", palette = "Spectral")+
  theme(
    legend.background = element_rect(
      fill = "lemonchiffon", 
      colour = "grey50", 
      size = 1
    )
  ) 
plot_bar(microBut, fill="Family")+facet_grid(condition~site, scales = "free")+theme(text = element_text(size=16))+geom_bar(stat="identity")+
  scale_fill_brewer(type = "seq", palette = "Spectral")+
  theme(
    legend.background = element_rect(
      fill = "lemonchiffon", 
      colour = "grey50", 
      size = 1
    )
  )
# Plotting PCoA
my.PCoA2 <- ordinate(microBut, "PCoA", "bray")
plot_ordination(microBut, my.PCoA2, type = "group", color = "condition", shape = "site")+theme(text = element_text(size=20))+ geom_point(size=5) + scale_color_manual(values=c("brown3", "steelblue","grey50"))
plot_ordination(microBut, my.PCoA2, type = "group", color = "condition")+ facet_wrap(~site, scales = "free_x", nrow = 1) + geom_point(size=5)+theme(text = element_text(size=20)) + scale_color_manual(values=c("brown3", "steelblue","grey50"))
plot_ordination(microBut, my.PCoA2, type = "group", color = "condition", shape = "site")+theme(text = element_text(size=20))+ geom_point(size=5)+ 
  geom_line() + scale_color_manual(values=c("brown3", "steelblue","grey50"))
plot_ordination(microBut, my.PCoA2, type = "group", color = "condition")+theme(text = element_text(size=20))+ geom_point(size=5)+ stat_ellipse(level=0.9)  + scale_color_manual(values=c("brown3", "steelblue","grey50"))
# PCoA plot using the unweighted UniFrac as distance
wunifrac_dist = phyloseq::distance(microBut, method="unifrac", weighted=F)
ordination = ordinate(microBut, method="PCoA", distance=wunifrac_dist)
plot_ordination(microBut, ordination, color="condition") + theme(aspect.ratio=1)+
  ggtitle("PCoA: unweigthed Unifrac")+geom_point(size=5)  + scale_color_manual(values=c("brown3", "steelblue","grey50"))
#Plot PCoA using the weighted UniFrac as distance
wunifrac_dist = phyloseq::distance(microBut, method="unifrac", weighted=T)
ordUF = ordinate(microBut, method="PCoA", distance=wunifrac_dist)
plot_ordination(microBut, ordUF, color = "condition") + 
  ggtitle("PCoA: weigthed Unifrac")+geom_point(size=5) + scale_color_manual(values=c("brown3", "steelblue","grey50"))
```  

## Relative abundance of butyrate-producing families
```{r}
phyBut <- normalize(phytaxBut, method = "relative")
# Plot taxonomy by group
plotbar(phyBut,level="Family", group="condition") + theme(axis.text.x = element_text(size = 16))
plotbar(phyBut,level="Family", group="site") + theme(axis.text.x = element_text(size = 16))

p1 <- phytaxBut
p1.gen <- aggregate_taxa(phytaxBut,"Genus")
x.d1 <- dominant_taxa(p1,level = "Genus", group="condition")
head(x.d1$dominant_overview)
tx.sum2 <- taxa_summary(p1, "Family")
?taxa_summary()
p1.rel <- microbiome::transform(p1, "compositional")

grp_abund1 <- get_group_abundances(p1.rel, 
                                  level = "Family", 
                                  group="condition",
                                  transform = "compositional")

mycols <- c("brown3", "steelblue","grey50")
# clean names 
grp_abund1$OTUID <- gsub("p__", "",grp_abund1$OTUID)
grp_abund1$OTUID <- ifelse(grp_abund1$OTUID == "", 
                          "Unclassified", grp_abund1$OTUID)
re <-reorder(grp_abund1$OTUID, grp_abund1$mean_abundance)
mean.plot1 <- grp_abund1 %>% # input data
  ggplot(aes(x=re, # reroder based on mean abundance
             y= mean_abundance,
             fill=condition)) + # x and y axis 
  geom_bar(     stat = "identity", 
                position=position_dodge()) + 
  scale_fill_manual("condition", values=mycols) + # manually specify colors
  theme_bw() + # add a widely used ggplot2 theme
  ylab("Mean Relative Abundance") + # label y axis
  xlab("Family") + # label x axis
  coord_flip() # rotate plot 
mean.plot1

pa <- aggregate_taxa(p1, "Family")
top_four <- top_taxa(pa, 4)
top_four
mycols <- c("brown3", "steelblue", "grey50")
p <- plot_listed_taxa(pa, top_four, 
                      group= "condition",
                      group.order = c("10%Fat","Oleate","60%Fat"),
                      group.colors = mycols,
                      add.violin = TRUE,
                      violin.opacity = 0.3,
                      dot.opacity = 0.25,
                      box.opacity = 0.25,
                      panel.arrange= "wrap")
comps <- make_pairs(sample_data(pa)$condition)
p <- p + stat_compare_means(
  comparisons = comps,
  label = "p.format",
  tip.length = 0.05,
  method = "wilcox.test")
print(p + ylab("Relative abundance") + scale_y_continuous(labels = scales::percent))
```  

## Normalization
```{r}
phy <- normalize(microBut, method = "relative")
# Plot taxonomy by group
plotbar(phy,level="Family", group="condition") + theme(axis.text.x = element_text(size = 16))
plotbar(phy,level="Family", group="site") + theme(axis.text.x = element_text(size = 16))
# Plot alpha diversity
plotalpha(microBut, group = "condition", color = c("brown3","grey50", "steelblue"))
plotalpha(microBut, group = "site")
# Test for significance between groups
beta_condition <-betatest(phy,group="condition")
beta_site <-betatest(phy,group="site")
# Biomarkers discovery and plotting
bio <- biomarker(microBut,group="condition")
plotmarker(bio,level="Genus")
# LefSe testing and plotting
lda3 <- ldamarker(microBut,group="condition")
lda3
write.csv(lda3, "lda3.csv", row.names = FALSE)
lda4 <- read.csv(file = 'lda4.csv')
plotLDA(lda4,group=c("10%Fat","60%Fat"),lda=5,pvalue=0.05)+theme(axis.text.y = element_text(size = 16))+theme(text = element_text(size=20))
plotLDA(lda4,group=c("10%Fat","Oleate"),lda=5,pvalue=0.05)+theme(axis.text.y = element_text(size = 16))+theme(text = element_text(size=20))
plotLDA(lda4,group=c("Oleate","60%Fat"),lda=5,pvalue=0.05) +theme(axis.text.y = element_text(size = 16))+theme(text = element_text(size=20))
```

## Modified difftest function from microbial package to test for significance
```{r, results='hide'}
# create difftest2 function
difftest2<-function(physeq,group,pvalue=0.05,padj=NULL,log2FC=0,gm_mean=TRUE,fitType="local",quiet=FALSE){
  if(!taxa_are_rows(physeq)){
    physeq<-t(physeq)
  }
  otu <- as(otu_table(physeq),"matrix")
  tax <- as.data.frame(as.matrix(tax_table(physeq)))
  colData<-as(sample_data(physeq),"data.frame")
  colData$condition<-colData[,group]
  contrasts<-levels(factor(unique(colData$condition)))[1:2]
  if(isTRUE(gm_mean)){
    countData<-round(otu, digits = 0)
  }else{
    countData<-round(otu, digits = 0)+1
  }
  dds <- DESeqDataSetFromMatrix(countData, colData, as.formula(~condition))
  if(isTRUE(gm_mean)){
    geoMeans = apply(counts(dds), 1, gm_mean)
    dds = estimateSizeFactors(dds, geoMeans = geoMeans)
  }
  dds <- DESeq(dds, fitType=fitType)
  res <- results(dds,contrast=c("condition",contrasts),cooksCutoff = FALSE)
  res_tax = cbind(as.data.frame(res), as.matrix(countData[rownames(res), ]))
  if(!is.null(padj)){
    pval<-padj
    sig <- rownames(subset(res,padj<pval&abs(log2FoldChange)>log2FC))
  }else{
    pval<-pvalue
    sig <- rownames(subset(res,pvalue<pval&abs(log2FoldChange)>log2FC))
  }
  res_tax$Significant<- "No"
  res_tax$Significant <- ifelse(rownames(res_tax) %in% sig, "Yes", "No")
  res_tax <- cbind(res_tax, tax[rownames(res),])
  return(as.data.frame(res_tax))
}
# Test for significant bacteria using microbial package
mic_res <- difftest2(microBut,group="condition", gm_mean=FALSE)
plotdiff(mic_res,level="Genus",padj=0.001,log2FC = 3,fontsize.y = 14)
```  

## Test for significant OTUs using DESeq2  
```{r}
sample_data(microBut)$condition <- as.factor(sample_data(microBut)$condition)
ds = phyloseq_to_deseq2(microBut, ~ condition)
ds = DESeq(ds)
alpha = 0.01
# compare Oleate and 60%Fat
res = results(ds, contrast=c("condition", "Oleate", "60%Fat"), alpha=alpha)
res = res[order(res$padj, na.last=NA), ]
res_sig = res[(res$padj < alpha), ]
res_sig
res_sig = cbind(as(res_sig, "data.frame"), as(tax_table(microBut)[rownames(res_sig), ], "matrix"))
ggplot(res_sig, aes(x=Genus, y=log2FoldChange, color=Phylum)) +
  geom_jitter(size=5, width = 0.2) + theme(text = element_text(size=16)) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
# compare 10%Fat and 60%Fat
res2 = results(ds, contrast=c("condition", "10%Fat", "60%Fat"), alpha=alpha)
res2 = res2[order(res2$padj, na.last=NA), ]
res_sig2 = res2[(res2$padj < alpha), ]
res_sig2
res_sig2 = cbind(as(res_sig2, "data.frame"), as(tax_table(microBut)[rownames(res_sig2), ], "matrix"))
ggplot(res_sig2, aes(x=Genus, y=log2FoldChange, color=Phylum)) +
  geom_jitter(size=5, width = 0.2) + theme(text = element_text(size=16)) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
# compare 10%Fat and Oleate
res3 = results(ds, contrast=c("condition", "10%Fat", "Oleate"), alpha=alpha)
res3 = res3[order(res3$padj, na.last=NA), ]
res_sig3 = res3[(res3$padj < alpha), ]
res_sig3
res_sig3 = cbind(as(res_sig3, "data.frame"), as(tax_table(microBut)[rownames(res_sig3), ], "matrix"))
ggplot(res_sig3, aes(x=Genus, y=log2FoldChange, color=Phylum)) +
  geom_jitter(size=5, width = 0.2) + theme(text = element_text(size=16)) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
```

## Plot heatmap  
```{r, results='hide'}
plot_taxa_heatmap(microBut,
                  subset.top = 20,
                  VariableA = "condition",
                  heatcolors = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
                  transformation = "log10"
)
```  

## Venn Diagram
```{r}
pseq<-microBut
table(meta(pseq)$condition, useNA = "always")
# convert to relative abundances
pseq.rel <- microbiome::transform(pseq, "compositional")
disease_states <- unique(as.character(meta(pseq.rel)$condition))
print(disease_states)
#Write a for loop to go through each of the disease_states one by one and combine identified core taxa into a list.
list_core <- c() # an empty object to store information

for (n in disease_states){ # for each variable n in DiseaseState
  #print(paste0("Identifying Core Taxa for ", n))
  
  ps.sub <- subset_samples(pseq.rel, condition == n) # Choose sample from DiseaseState by n
  
  core_m <- core_members(ps.sub, # ps.sub is phyloseq selected with only samples from g 
                         detection = 0.001, # 0.001 in atleast 90% samples 
                         prevalence = 0.75)
  print(paste0("No. of core taxa in ", n, " : ", length(core_m))) # print core taxa identified in each DiseaseState.
  list_core[[n]] <- core_m # add to a list core taxa for each group.
  #print(list_core)
}
print(list_core)
# Specify colors and plot venn
# supplying colors in the order they appear in list_core
mycols <- c("brown3", "grey50", "steelblue")
venn.diagram(list_core,fill = mycols, filename = 'But_venn_diagramm.png',
             output=TRUE)
```  
![Butyrate_Venn](But_venn_diagramm.png)

## Plotting top 4 genus  
```{r, warning=FALSE}
pa <- aggregate_taxa(microBut, "Genus")
top_four <- top_taxa(pa, 4)
top_four
mycols <- c("brown3", "steelblue", "grey50")
p <- plot_listed_taxa(pa, top_four, 
                      group= "condition",
                      group.order = c("10%Fat","Oleate","60%Fat"),
                      group.colors = mycols,
                      add.violin = TRUE,
                      violin.opacity = 0.3,
                      dot.opacity = 0.25,
                      box.opacity = 0.25,
                      panel.arrange= "wrap")
comps <- make_pairs(sample_data(pa)$condition)
p <- p + stat_compare_means(
  comparisons = comps,
  label = "p.format",
  tip.length = 0.05,
  method = "wilcox.test")
print(p + ylab("Relative abundance") + scale_y_continuous(labels = scales::percent))
```

## Get taxa summary by group(s)
```{r}
p0 <- microBut
p0.gen <- aggregate_taxa(microBut,"Genus")
x.d <- dominant_taxa(p0,level = "Genus", group="condition")
head(x.d$dominant_overview)
tx.sum1 <- taxa_summary(p0, "Phylum")
p0.rel <- microbiome::transform(p0, "compositional")

grp_abund <- get_group_abundances(p0.rel, 
                                  level = "Family", 
                                  group="condition",
                                  transform = "compositional")
mycols <- c("brown3", "steelblue","grey50")
# clean names 
grp_abund$OTUID <- gsub("p__", "",grp_abund$OTUID)
grp_abund$OTUID <- ifelse(grp_abund$OTUID == "", 
                          "Unclassified", grp_abund$OTUID)
mean.plot <- grp_abund %>% # input data
  ggplot(aes(x= reorder(OTUID, mean_abundance), # reroder based on mean abundance
             y= mean_abundance,
             fill=condition)) + # x and y axis 
  geom_bar(     stat = "identity", 
                position=position_dodge()) + 
  scale_fill_manual("condition", values=mycols) + # manually specify colors
  theme_bw() + # add a widely used ggplot2 theme
  ylab("Mean Relative Abundance") + # label y axis
  xlab("Family") + # label x axis
  coord_flip() # rotate plot 
mean.plot
```  

## Plotting density of reads per sample  
```{r}
p0 <- microBut
print_ps(p0)
kable(head(tax_table(p0)))
# reduce size for example
ps0 <- core(p0, detection = 10, prevalence = 20 / 100)
# Add a prefix to taxa labels
ps0.f2 <- format_to_besthit(ps0, prefix = "MyBug-")
kable(head(tax_table(ps0.f2))[3:6])
p <- plot_read_distribution(ps0.f2, groups = "condition", 
                            plot.type = "density")
print(p + theme_biome_utils())
ps0 <- core(p0, detection = 10, prevalence = 20 / 100)
pseq_df <- phy_to_ldf(ps0, transform.counts = NULL)
kable(head(pseq_df))
# Plot Density of phyla
pseq <- microBut
# check 10%Fat
control_ps <- subset_samples(pseq, condition=="10%Fat")
p_hc <- taxa_distribution(control_ps) + 
  theme_biome_utils() + 
  labs(title = "10%Fat")

# check Oleate
oleate_ps <- subset_samples(pseq, condition=="Oleate")
p_oleate <- taxa_distribution(oleate_ps) + 
  theme_biome_utils() + 
  labs(title = "Oleate")

# check 60%Fat
HFD_ps <- subset_samples(pseq, condition=="60%Fat")
p_hfd <- taxa_distribution(HFD_ps) + 
  theme_biome_utils() + 
  labs(title = "60%Fat")

# harnessing the power of patchwork
p_hc / p_oleate / p_hfd + plot_layout(guides = "collect") + 
  plot_annotation(tag_levels = "A")
```

# Session Info  
```{r}
sessionInfo()
```